runtime/src/main/java/org/corfudb/runtime/collections/CorfuTable.java

```diff
@@ -0,0 +1,642 @@
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Mutator(name = "putAll",
+            undoFunction = "undoPutAll",
+            undoRecordFunction = "undoPutAllRecord",
+            conflictParameterFunction = "putAllConflictFunction")
+    public void putAll(@Nonnull Map<? extends K, ? extends V> m) {
+        // If we have no index functions, then just directly put all
+        if (indexFunctions.isEmpty()) {
+            mainMap.putAll(m);
+        }
+        // Otherwise we must update all secondary indexes
+        // TODO: Do this in parallel (need to acquire update locks, potentially)
+        m.entrySet().stream()
+                .forEach(e -> {
+                    V previous = mainMap.put(e.getKey(), e.getValue());
+                    unmapSecondaryIndexes(e.getKey(), previous);
+                    mapSecondaryIndexes(e.getKey(), e.getValue());
+                });
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Mutator(name = "clear", reset = true)
+    public void clear() {
+        mainMap.clear();
+        indexMap.values().parallelStream().forEach(m -> m.clear());
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Accessor
+    public @Nonnull Set<K> keySet() {
+        return ImmutableSet.copyOf(mainMap.keySet());
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Accessor
+    public @Nonnull Collection<V> values() {
+        return ImmutableList.copyOf(mainMap.values());
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Accessor
+    public @Nonnull Set<Entry<K, V>> entrySet() {
+        return ImmutableSet.copyOf(mainMap.entrySet());
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Accessor
+    public V getOrDefault(Object key, V defaultValue) {
+        return mainMap.getOrDefault(key, defaultValue);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Accessor
+    public void forEach(BiConsumer<? super K, ? super V> action) {
+        mainMap.forEach(action);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @TransactionalMethod
+    public void replaceAll(BiFunction<? super K, ? super V, ? extends V> function) {
+        Objects.requireNonNull(function);
+        for (Map.Entry<K, V> entry : entrySet()) {
+            K k;
+            V v;
+            try {
+                k = entry.getKey();
+                v = entry.getValue();
+            } catch (IllegalStateException ise) {
+                // this usually means the entry is no longer in the map.
+                throw new ConcurrentModificationException(ise);
+            }
+
+            // ise thrown from function is not a cme.
+            v = function.apply(k, v);
+
+            try {
+                entry.setValue(v);
+            } catch (IllegalStateException ise) {
+                // this usually means the entry is no longer in the map.
+                throw new ConcurrentModificationException(ise);
+            }
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @TransactionalMethod
```

Comment:
Does this mean it is under a lock ? If so then it is fine because I think CAS semantics hold still. Could you please verify and also put comments about this. Otherwise looking at code it seems like get , compare and put are different statements and context switching can happen in between them .